sql/analyzer/resolve_variables.go

```diff
@@ -23,109 +23,116 @@ import (
 	"github.com/dolthub/go-mysql-server/sql"
 	"github.com/dolthub/go-mysql-server/sql/expression"
 	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
 )
 
 // resolveVariables replaces UnresolvedColumn which are variables with their literal values
-func resolveVariables(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+func resolveVariables(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, transform.TreeIdentity, error) {
 	span, ctx := ctx.Span("resolve_variables")
 	defer span.Finish()
 
-	return plan.TransformUp(n, func(n sql.Node) (sql.Node, error) {
-		if n.Resolved() {
-			return n, nil
+	return transform.Node(n, func(node sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		if node.Resolved() {
+			return node, transform.SameTree, nil
 		}
 
-		resolveVars := func(e sql.Expression) (sql.Expression, error) {
+		resolveVars := func(e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
 			uc, ok := e.(column)
 			if !ok || e.Resolved() {
-				return e, nil
+				return e, transform.SameTree, nil
 			}
 
-			expr, ok, err := resolveSystemOrUserVariable(ctx, a, uc)
+			expr, same, err := resolveSystemOrUserVariable(ctx, a, uc)
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
 			}
-			if ok {
-				return expr, nil
+			if same {
+				return e, transform.SameTree, nil
 			}
-
-			return e, nil
+			return expr, transform.NewTree, nil
 		}
 
 		// Set nodes need to resolve the right-hand side of an expression only
-		if n, ok := n.(*plan.Set); ok {
-			return plan.TransformExpressionsUp(n, func(e sql.Expression) (sql.Expression, error) {
+		if n, ok := node.(*plan.Set); ok {
+			return transform.NodeExprsWithNode(n, func(_ sql.Node, e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
 				sf, ok := e.(*expression.SetField)
 				if !ok {
-					return e, nil
+					return e, transform.SameTree, nil
 				}
 
-				nr, err := expression.TransformUp(sf.Right, resolveVars)
+				nr, same, err := transform.Expr(sf.Right, resolveVars)
 				if err != nil {
-					return nil, err
+					return nil, transform.SameTree, err
 				}
 
-				return sf.WithChildren(sf.Left, nr)
+				if same {
+					return e, transform.SameTree, nil
+				}
+				e, err = sf.WithChildren(sf.Left, nr)
+				if err != nil {
+					return nil, transform.SameTree, err
+				}
+				return e, transform.NewTree, nil
 			})
 		}
 
-		return plan.TransformExpressionsUp(n, resolveVars)
+		return transform.OneNodeExpressions(node, resolveVars)
```

Comment:
This changed the semantics (from the entire tree to the single node). On purpose?

Reply:
We are inside a tree traversal and are already hitting every node, there's no need to duplicate